Jwt security token

TEdwardJ opened the pull request.

TEdwardJ committed:
JWT Security:
 - JWT logout functionality added;

Greengrer committed:
Added access for 'https://react-wax-deals.herokuapp.com'.

Greengrer committed:
Added access for 'https://react-wax-deals.herokuapp.com' second try.

Greengrer committed:
Added access for 'https://react-wax-deals.herokuapp.com' to all contollers now third try.

Greengrer committed:
Fixed deejay parser that pulled equipment along with vinyls from store, not it's only vinyls.

TEdwardJ committed:
JWT Security, Registration + Confirmation:
 - general CORS settings added;
 - decommission of old registration and confirmation controllers, but not elimination yet;
 - RestSignupController added in order to cover registration and confirmation functionality;
 - Tests for RestSignupController added;
 - UserDetails implementation JwtUser has got enabled/disabled flag that Spring Security can handle;
 - JwtLoginControllerITest - test for login of non-activated user is added;
 -

TEdwardJ committed:
JWT Security:
 - merging of changes made in parallel;

TEdwardJ committed:
JWT Security:
 - ControllerResponseUtils class is created to serve the purpose of preparing controllers response;
 - JwtLoginController simplified in order to contain only one service call;

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/JwtUser.java:
```diff
@@ -12,6 +12,8 @@
     private String username;
     private String password;
+
+    private boolean isEnabled;
```
should be "enabled", "isEnabled" is good practice for getter method's name

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/JwtUser.java:
```diff
@@ -58,6 +60,11 @@ public boolean isCredentialsNonExpired() {
     }
+
+    public void setEnabled(boolean enabled) {
+        isEnabled = enabled;
```
this.enabled = enabled

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+
+@Service
+@RequiredArgsConstructor
+public class InMemoryLogoutTokenStorageService implements LogoutTokenStorageService {
```
Either "InMemoryLogoutTokenStorage" or "InMemoryLogoutTokenService"

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+    @PostConstruct
+    public void initCache() {
+        logoutTokensCache = Caffeine.newBuilder()
+                .expireAfter(new Expiry<String, LocalDateTime>() {
```
encapsulate this into private method and use this::methodName instead

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+                    private ZoneOffset offset = ZoneOffset.ofTotalSeconds(0);
+
+                    @Override
+                    public long expireAfterCreate(@NonNull String s, @NonNull LocalDateTime localDateTime, long l) {
```
bad parameters names

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+                    }
+
+                    @Override
+                    public long expireAfterUpdate(@NonNull String s, @NonNull LocalDateTime localDateTime, long l, @NonNegative long l1) {
```
bad parameters names

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+                    }
+
+                    @Override
+                    public long expireAfterRead(@NonNull String s, @NonNull LocalDateTime localDateTime, long l, @NonNegative long l1) {
```
bad parameters names

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenStorageService.java:
```diff
@@ -0,0 +1,68 @@
+                    }
+
+                    private long getDifference(@NonNull LocalDateTime localDateTime) {
+                        return 1_000_000_000L * (localDateTime.toEpochSecond(offset) - LocalDateTime.now().toEpochSecond(offset));
```
magic number

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/LogoutHandlerService.java:
```diff
@@ -0,0 +1,34 @@
+
+
+@Service
+public class LogoutHandlerService implements LogoutHandler {
```
Either "LogoutHandler" or "LogoutService"

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/LogoutHandlerService.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public void logout(HttpServletRequest request, HttpServletResponse response, Authentication authentication) {
+
```
obsolete extra line

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/LogoutHandlerService.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public void logout(HttpServletRequest request, HttpServletResponse response, Authentication authentication) {
+
+        //extract token
```
Please remove these comments and never use them
Your code should be self descriptive

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/LogoutHandlerService.java:
```diff
@@ -0,0 +1,34 @@
+        //extract token
+        String token = jwtService.extractToken(request);
+        //check validity
+        if (jwtService.validateToken(token)) {
```
jwtService.isTokenValid()

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/SecurityConfiguration.java:
```diff
@@ -50,14 +56,26 @@ public void configure(WebSecurity web) throws Exception {
-                .formLogin().loginPage("/signIn").permitAll().usernameParameter("email").defaultSuccessUrl("/profile");/*.successHandler()*///;
+                .formLogin().loginPage("/signIn").permitAll().usernameParameter("email").defaultSuccessUrl("/profile")
+                .and()
+                .logout().addLogoutHandler(logoutHandlerService).logoutSuccessUrl("/successlogout");
```
Seems like you do not have "failedLogout" so just "/logout" is better than "successLogout"

TEdwardJ replied:
we have /logout set by Spring Security

Rooman replied:
do we actually need a success URL for that functionality? We are REST service

TEdwardJ replied:
yes, we do...this controller method returns the appropriate answer after Spring Security did its logout magic.
I see shortcoming it this decision, since the endpoint is exposed and hence can be requested from outside but this call will not make any sense.
Could you please advise whether there is a better way to do the same thing?
 TIA

littlepony99 replied:
issue was created

https://app.clickup.com/t/u3xqgm

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/TokenCacheConfig.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Bean
+    public Caffeine caffeineConfig() {
+        return Caffeine
```
Absolutely readable in one line. If you dont have opposite code convetion on the project ofc

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/impl/DefaultSecurityService.java:
```diff
@@ -59,8 +59,7 @@ public User createUserWithHashedPassword(String email, char[] password) {
         if (user != null) {
-            isSame = (user.getPassword().equals(hashPassword(password,
-                    Base64.getDecoder().decode(user.getSalt()), user.getIterations())));
+            isSame = encoder.matches(new String(password), user.getPassword());
```
"matched" or "same", "is*" is used for methods

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/service/JwtService.java:
```diff
@@ -16,4 +19,9 @@
 
+    LocalDateTime getExpirationDate(String token);
+
+    Map<String, Object> checkTokenAndReturnCredentialsMap(String token);
```
awful field name. please reconsider it

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/service/JwtService.java:
```diff
@@ -16,4 +19,9 @@
+
+    Map<String, Object> checkTokenAndReturnCredentialsMap(String token);
+
+    Map<String, Object> authenticateAndReturnCredentialsMap(LoginRequest loginRequest);
```
awful field name. please reconsider it

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/service/JwtTokenProvider.java:
```diff
@@ -16,24 +23,40 @@
-    @Value("${jwt.token.expired:120000}")
-    private int validityInMilliseconds;
+    @Autowired
+    public void setTokenStorageService(LogoutTokenStorageService tokenStorageService) {
```
Why setter autowiring not field one ?

TEdwardJ replied:
it is to avoid cyclic references in the app`s Spring configuration.
Could you please advise is it a good practice to use autowired setters?

hemiSRT8 replied:
According to spring documentation: 
`Since you can mix constructor-based and setter-based DI, it is a good rule of thumb to use constructors for mandatory dependencies and setter methods or configuration methods for optional dependencies`

littlepony99 replied:
we will work with it in the next sprint

https://app.clickup.com/t/u3xuc6

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultEmailConfirmationService.java:
```diff
@@ -52,12 +52,20 @@
 
+    @Override
+    public UUID generateToken() {
+
```
obsolete extra line

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java:
```diff
@@ -133,9 +133,9 @@ public void signInCheck(UserInfoRequest userProfileInfo) {
         if (!securityService.validateIfPasswordMatches(userToCheckAgainst, userProfileInfo.getPassword().toCharArray())) {
             throw new RuntimeException(UserErrors.WRONG_CREDENTIALS_ERROR.getMessage());
         }
-        if (!userToCheckAgainst.getStatus()) {
```
What is the purpose to comment this code? If it is not needed - remove it and you could view it in git history if you need it
YAGNI principle

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/util/ControllerResponseUtils.java:
```diff
@@ -0,0 +1,29 @@
+
+    private static UserMapper userMapper = new DefaultUserMapper();
+
+    public static Map<String, String> getStatusInfoMap(String code, String s) {
```
bad parameter name

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/EmailConfirmationController.java:
```diff
@@ -1,25 +1,28 @@
 @RequiredArgsConstructor
-@Controller
-@RequestMapping("/emailConfirmation")
+/*@Controller
```
YAGNI  - remove commented code at all and use git history

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/JwtLoginController.java:
```diff
@@ -1,46 +1,43 @@
+
+    @GetMapping(value = "/successlogout", produces = {MediaType.APPLICATION_JSON_VALUE})
+    public ResponseEntity<Map<String, Object>> logout() {
+        Map<String, Object> response = new HashMap<>(getSuccessStatusInfoMap());
```
obsolete variable declaration

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+@Slf4j
+@RequiredArgsConstructor
+@CrossOrigin(origins = {"http://localhost:3000", "http://react-wax-deals.herokuapp.com"})
+public class RestSignupController {
```
Do you have not "REST" signup controllers ? If not - just "SignupController"

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+
+
+    @PostMapping("/signUp")
+    public ResponseEntity<Map<String, Object>> signUpUser(@RequestBody UserInfoRequest userProfileInfo) {
```
do you sign up only users? if so - just "signUp"

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+        } catch (Exception e) {
+            log.error("Error during registration", e);
+            responseMap.putAll(ControllerResponseUtils.getStatusInfoMap("1", e.getMessage()));
+            ResponseEntity<Map<String, Object>> response = new ResponseEntity<>(responseMap, HttpStatus.BAD_REQUEST);
```
obsolete variable declaration

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+            responseMap.putAll(ControllerResponseUtils.getUserCredentialsMap(jwtService.createToken(confirmedUser.getUsername(), confirmedUser.getAuthorities()), user));
+            responseMap.putAll(ControllerResponseUtils.getStatusInfoMap("0", ""));
+            log.debug("Set response status to {'status':{}}", HttpStatus.OK);
+            ResponseEntity<Map<String, Object>> response = new ResponseEntity(responseMap, HttpStatus.OK);
```
obsolete variable declaration

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+        } catch (Exception e){
+            log.error("Error during confirmation ", e);
+            responseMap.putAll(ControllerResponseUtils.getStatusInfoMap("1", e.getMessage()));
+            ResponseEntity<Map<String, Object>> response = new ResponseEntity<>(responseMap, HttpStatus.BAD_REQUEST);
```
obsolete variable declaration

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+            return response;
+        }
+    }
+
```
A lot of obsolete extra lines

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/RestSignupController.java:
```diff
@@ -0,0 +1,87 @@
+                .map(foundConfirmationToken -> foundConfirmationToken.getUserId())
+                .flatMap(userFromToken -> userService.findById(userFromToken))
+                .get();
+        return user;
```
obsolete variable declaration

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/SignInController.java:
```diff
@@ -43,7 +42,7 @@ public String getSignInPage(@SessionAttribute(value = "user", required = false)
     }
 
-    @PostMapping("/emailConfirmation")
+    //@PostMapping("/emailConfirmation")
```
YAGNI

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/controller/SignInController.java:
```diff
@@ -9,15 +9,14 @@
 
 @Slf4j
-@Controller
+//@Controller
```
YAGNI

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/web/handler/ApiExceptionHandler.java:
```diff
@@ -6,7 +6,7 @@
 import static org.springframework.http.HttpStatus.BAD_REQUEST;
 
-@ControllerAdvice
+//@ControllerAdvice
```
YAGNI

hemiSRT8 commented on src/main/resources/application.properties:
```diff
@@ -1,15 +1,21 @@
 FILE_NAME=vinyl.log
 FILE_NAME_PATTERN=vinyl-%d{yyyy-MM-dd}.%i.log
-FILTER_LEVEL=info
+FILTER_LEVEL=debug
```
sure?

TEdwardJ committed:
JWT Security pull request reviews:
 - obsolete lines;
 - obsolete variables;
 - method naming changes;
 - class naming changes;

Rooman commented on src/main/java/com/vinylteam/vinyl/ApplicationCommandLineRunner.java:
```diff
@@ -0,0 +1,30 @@
+
+//@Component
+@RequiredArgsConstructor
+public class ApplicationCommandLineRunner implements CommandLineRunner {
```
Why do we need this class

Rooman commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenService.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.concurrent.TimeUnit;
+
+@Service
+@RequiredArgsConstructor
```
why do we need this annotation?

Rooman commented on src/main/java/com/vinylteam/vinyl/security/LogoutService.java:
```diff
@@ -0,0 +1,29 @@
+public class LogoutService implements LogoutHandler {
+
+    @Autowired
+    private JwtService jwtService;
```
don't use field injection

Rooman commented on src/main/java/com/vinylteam/vinyl/security/SecurityConfiguration.java:
```diff
@@ -50,14 +53,26 @@ public void configure(WebSecurity web) throws Exception {
+    @Bean
+    public CorsConfigurationSource corsConfigurationSource() {
+        CorsConfiguration configuration = new CorsConfiguration();
+        configuration.setAllowedOrigins(Arrays.asList("http://localhost:3000", "http://react-wax-deals.herokuapp.com"));
```
application propertes?

TarasOrlovskyi replied:
Please, change it 'http://react-wax-deals.herokuapp.com' to 'https://react-wax-deals.herokuapp.com'. I mean that request will be come from 'https'

Rooman commented on src/main/java/com/vinylteam/vinyl/service/EmailConfirmationService.java:
```diff
@@ -12,6 +13,8 @@
 
     ConfirmationToken addByUserId(long userId);
 
+    UUID generateToken();
```
For me it's not clear what kind of token emailService should generate

Rooman commented on src/main/java/com/vinylteam/vinyl/service/JwtService.java:
```diff
@@ -1,19 +1,28 @@
 
+    LocalDateTime getExpirationDate(String token);
+
+    Map<String, Object> checkToken(String token);
```
check for what?

Rooman commented on src/main/java/com/vinylteam/vinyl/service/JwtService.java:
```diff
@@ -1,19 +1,28 @@
+
+    Map<String, Object> checkToken(String token);
+
+    Map<String, Object> authenticateByRequest(LoginRequest loginRequest);
```
Do we sure that Map is the best return parameter for this method?

Rooman commented on src/main/java/com/vinylteam/vinyl/util/ControllerResponseUtils.java:
```diff
@@ -0,0 +1,29 @@
+
+public class ControllerResponseUtils {
+
+    private static UserMapper userMapper = new DefaultUserMapper();
```
static final?

Rooman commented on src/main/java/com/vinylteam/vinyl/util/impl/DeejayDeParser.java:
```diff
@@ -18,7 +18,11 @@
     protected static final String BASE_LINK = "https://www.deejay.de";
-    protected static final String START_PAGE_LINK = BASE_LINK + "/m_All/sm_News/stock_1/perpage_160";
+    protected static final String GENRE_HOUSE = "/m_House/sm_News/stock_1/perpage_160";
+    protected static final String GENRE_TECHNO = "/m_Techno/sm_News/stock_1/perpage_160";
```
protected?

TarasOrlovskyi commented on src/main/java/com/vinylteam/vinyl/ApplicationCommandLineRunner.java:
```diff
@@ -0,0 +1,30 @@
+                .password("TestPassword24")
+                .passwordConfirmation("TestPassword24")
+                .build();
+        //var user = securityService.createUserWithHashedPassword("test_user@gmail.com","superPAssword".toCharArray());
+        userService.register(request);
+        //updater.updateUniqueVinylsRewriteOffers();
```
you should to delete comments

TarasOrlovskyi commented on src/main/java/com/vinylteam/vinyl/security/SecurityConfiguration.java:
```diff
@@ -50,14 +53,26 @@ public void configure(WebSecurity web) throws Exception {
+    public CorsConfigurationSource corsConfigurationSource() {
+        CorsConfiguration configuration = new CorsConfiguration();
+        configuration.setAllowedOrigins(Arrays.asList("http://localhost:3000", "http://react-wax-deals.herokuapp.com"));
+        configuration.setAllowedMethods(Arrays.asList("GET", "POST"));
```
we will have PUT and DELETE request also

TarasOrlovskyi commented on src/main/java/com/vinylteam/vinyl/service/JwtTokenProvider.java:
```diff
@@ -16,53 +23,74 @@
-    @Value("${jwt.token.expired:120000}")
+    private LogoutTokenStorageService tokenStorageService;
+
+    @Value("${jwt.token.expired:300000}")
```
you should transfer it into application.properties. 300000 - hard to read. Do you mean 5 min? I think it very small period. Maybe it only for demo

TEdwardJ replied:
It is just default value in case Spring has not found the property in available property sources.
Nonetheless, it is really hard to read that value, so the property is handled now as number of seconds.
Moreover, the default is set to 10 minutes

TarasOrlovskyi commented on src/main/java/com/vinylteam/vinyl/web/controller/AboutController.java:
```diff
@@ -11,7 +11,7 @@
 @Controller
 @RequestMapping(path = "/about")
-@CrossOrigin(origins = {"http://localhost:3000", "http://react-wax-deals.herokuapp.com"})
+@CrossOrigin(origins = {"http://localhost:3000", "https://react-wax-deals.herokuapp.com"})
```
if you have SecurityConfiguration.java this configure is unnecessary. I mean everywhere in project

TarasOrlovskyi commented on src/main/java/com/vinylteam/vinyl/web/controller/SignupController.java:
```diff
@@ -0,0 +1,80 @@
+@RestController
+@Slf4j
+@RequiredArgsConstructor
+@CrossOrigin(origins = {"http://localhost:3000", "http://react-wax-deals.herokuapp.com"})
```
not 'http', but 'https'. But I think that now it unnecessary

TEdwardJ committed:
JWT Security:
 - rework based on pull request review'

TEdwardJ committed:
JWT Security:
 - unused imports eliminations;

TEdwardJ committed:
JWT Security:
 - elimination of redundant variable settings;
